### PR TITLE
nixos/github-runner: init at v2.277.1

### PR DESCRIPTION
**Pull request by veehaitch**
###### Motivation for this change

#101426

Adds [GitHub Actions Runner](https://github.com/actions/runner) to nixpkgs. Only `x86_64-linux` is supported right now. The .NET tests are run in the check phase and pass (apart from some online tests).

Required quite a few source patches to make it work with Nix:
- Tries to obtain the hash of the latest Git commit. Patched out and replaced by passing the `rev` from `fetchFromGitHub` to the property `GitInfoCommitHash` directly.
- Updates itself if a new release is detected. This cannot be disabled (see https://github.com/actions/runner/issues/485), hence, it's patched out.
- Assumes that its root is writable which is not the case for Nix as it resolves to the read-only Nix store. Introduced an environment variable `RUNNER_ROOT` which allows changing the root directory.
- Tries to install a systemd service when registering a new runner. Also patched out.
- The original release also includes Node. Replaced by a symlink to Nix's Node.

The output derivation tries to resemble the structure of the upstream releases where it makes sense. Also, the scripts `config.sh` and `run.sh` are fully usable.

Also adds a NixOS module:
- If any of the options `name`, `url`, `runnerGroup`, `extraLabels`, `tokenFile` or its content changes, a registration is triggered. 
- Relies on systemd life cycle management and isolation where possible.

Thanks @Trundle for helping out 🙏🏻 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/continuous-integration/github-runner/default.nix [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    # Disable specific tests
+    substituteInPlace src/dir.proj \
+      --replace 'dotnet test Test/Test.csproj' \
+                "dotnet test Test/Test.csproj --filter '${testFilterXml}'"
+
+    # Fix FHS path
+    substituteInPlace src/Test/L0/Util/IOUtilL0.cs \
+      --replace '/bin/ln' \
+                '${coreutils}/bin/ln'
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/github-runner/default.nix:126]:
```suggestion
      --replace '/bin/ln' '${coreutils}/bin/ln'
```

File: pkgs/development/tools/continuous-integration/github-runner/default.nix [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    ./patches/dont-install-systemd-service.patch
+    # Don't try to self-update runner (cannot be disabled, see https://github.com/actions/runner/issues/485)
+    ./patches/ignore-self-update.patch
+  ];
+
+  postPatch = ''
+    # Relax the version requirement
+    substituteInPlace src/global.json \
+      --replace '3.1.302' \
+                '${dotnetSdk.version}'
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/github-runner/default.nix:116]:
```suggestion
      --replace '3.1.302' '${dotnetSdk.version}'
```

File: pkgs/development/tools/continuous-integration/github-runner/default.nix [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+, kerberos
+, lib
+, linkFarm
+, lttng-ust
+, makeWrapper
+, nodejs-12_x
+, openssl
+, stdenv
+, zlib
+, ...
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/github-runner/default.nix:20]:
```suggestion
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116775` at 73456884 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/315c513d4e9588aff067ce9d3373d79a">github-runner</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/continuous-integration/github-runner/default.nix:174:3:
```
    |
174 |   checkPhase = ''
    |   ^
```
</li>
  </ul>
</details>

File: nixos/modules/services/continuous-integration/github-runner.nix [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+    name = mkOption {
+      type = types.nullOr types.str;
+      description = ''
+        Name of the runner to configure. Defaults to the hostname.
+
+        Changing this option triggers a new runner registration.
+      '';
+      example = "nixos";
+      default = null;
```
Comment (kfollesdal) [inline nixos/modules/services/continuous-integration/github-runner.nix:56]: I have testet your PR, work fine so fare :-) But get following error if I do not set option `name`. 
```
error: cannot coerce null to a string

       at /nix/store/rfxj941pv6yf3p2rikhx8wsw3y1q7m34-source/nixos/modules/services/continuous-integration/github-runner.nix:6:18:

            5|   svcName = "github-runner";
            6|   systemdUser = "${svcName}-${cfg.name}";
             |                  ^
            7|   systemdDir = "${svcName}/${cfg.name}";
```
Looks like hostname is not set as default as described. But maybe I have done something wrong. But after setting name it works fine for me.

Comment (veehaitch) [reply nixos/modules/services/continuous-integration/github-runner.nix]: Good catch, thanks! Hopefully this is fixed in 9dcd2ec516fcf30f1c34afe4122a82306da5ef62.

**Comment by domenkozar** [comment]
@veehaitch is this ready?

**Comment by veehaitch** [comment]
> @veehaitch is this ready?

I'd say: yes! I've been using the runner without any issues since opening this PR. 

Obviously, there's room for improvement. Mainly support for multiple runner definitions (currently, we can achieve this with containers) and runner registration using a personal access token. I'd suggest, however, to implement those improvements in follow-ups. This PR is pretty big already.

**Comment by zimbatm** [comment]
sweet!

**Comment by mweinelt** [comment]
@zimbatm This causes an eval error, because the kerberos attribute isn't pased.

> anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/pkgs/development/tools/continuous-integration/github-runner/default.nix:1:1 called without required argument 'kerberos', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-2/lib/customisation.nix:69:16

https://gist.github.com/GrahamcOfBorg/2f92163fa71808194f67a2b60974f5a3

**Comment by domenkozar** [comment]
How come CI didn't catch that?

**Comment by veehaitch** [comment]
https://github.com/NixOS/nixpkgs/commit/dcb501f99325 wasn't included yet in this branch's base

**Comment by SuperSandro2000** [comment]
> How come CI didn't catch that?

The ofborg results are a week old I think.

**Comment by dhess** [comment]
Thanks, this is fantastic!

**Pull request merged**